Libplanet/Store/DefaultStore.cs

```diff
@@ -246,174 +271,237 @@ public override IEnumerable<TxId> IterateStagedTransactionIds()
-            byte[] bytes;
-
-            _txLock.EnterUpgradeableReadLock();
-            try
+            if (_txCache.TryGetValue(txid, out object cachedTx))
             {
-                LiteFileInfo file = _db.FileStorage.FindById(fileId);
-                if (file is null)
-                {
-                    return null;
-                }
-
-                using (var stream = new MemoryStream())
-                {
-                    DownloadFile(file, stream);
+                return (Transaction<T>)cachedTx;
+            }
 
-                    bytes = stream.ToArray();
-                    if (bytes.Length != file.Length || bytes.Length < 1)
-                    {
-                        _logger.Warning(
-                            "The data file for the transaction {TxId} seems corrupted; " +
-                            "it will be treated nonexistent and removed at all.",
-                            txid
-                        );
-                        _txLock.EnterWriteLock();
-                        try
-                        {
-                            _db.FileStorage.Delete(fileId);
-                        }
-                        finally
-                        {
-                            _txLock.ExitWriteLock();
-                        }
+            UPath path = TxPath(txid);
+            if (!_txs.FileExists(path))
+            {
+                return null;
+            }
 
-                        return null;
-                    }
-                }
+            byte[] bytes;
+            try
+            {
+                bytes = _txs.ReadAllBytes(path);
             }
-            finally
+            catch (FileNotFoundException)
             {
-                _txLock.ExitUpgradeableReadLock();
+                return null;
             }
 
-            return Transaction<T>.FromBencodex(bytes);
+            Transaction<T> tx = Transaction<T>.FromBencodex(bytes);
+            _txCache.AddOrUpdate(txid, tx);
+            return tx;
         }
 
         /// <inheritdoc/>
         public override void PutTransaction<T>(Transaction<T> tx)
         {
-            string fileId = TxFileId(tx.Id);
-            string filename = tx.Id.ToHex();
-            byte[] txBytes = tx.ToBencodex(true);
-            _txLock.EnterUpgradeableReadLock();
-            try
+            if (_txCache.ContainsKey(tx.Id))
             {
-                LiteFileInfo file = _db.FileStorage.FindById(fileId);
-                if (file is LiteFileInfo)
-                {
-                    // No-op if already exists.
-                    return;
-                }
+                return;
+            }
+
+            byte[] txBytes = tx.ToBencodex(true);
+            UPath path = TxPath(tx.Id);
+            UPath dirPath = path.GetDirectory();
+            CreateDirectoryRecursively(_txs, dirPath);
 
-                _txLock.EnterWriteLock();
+            if (_root is MemoryFileSystem)
```

Comment:
How about unifying path on `MemoeryFileSystem` and `SubFileSystem` to omit type checking?

Reply:
Actually, this is not for dealing with paths.  The current way to update files atomically, writing things into a temporary file and then renaming it to the final destination, does not work on `MemoryFileSystem`, so this branch puporses to work around that.

Reply:
Anyway, I'm going to try doing my best to avoid this kind of type query, which is indeed a code smell…

Reply:
Fixed: aa1744d92215dfd48dae119ade83dbad6bf6bb7e.